Userland/Shell/Builtin.cpp

```diff
@@ -1781,6 +1782,97 @@ ErrorOr<int> Shell::builtin_argsparser_parse(Main::Arguments arguments)
     return 0;
 }
 
+ErrorOr<int> Shell::builtin_read(Main::Arguments arguments)
+{
+    if (!m_in_posix_mode) {
+        raise_error(ShellError::EvaluatedSyntaxError, "read: POSIX builtin used in non-POSIX mode");
+        return 1;
+    }
+
+    bool no_escape = false;
+    Vector<DeprecatedString> variables;
+
+    Core::ArgsParser parser;
+    parser.add_option(no_escape, "Do not interpret backslash escapes", "no-escape", 'r');
+    parser.add_positional_argument(variables, "Variables to read into", "variable", Core::ArgsParser::Required::Yes);
+
+    if (!parser.parse(arguments, Core::ArgsParser::FailureBehavior::Ignore))
+        return 1;
+
+    auto split_by_any_of = " \t\n"_short_string;
+    auto split_by_variable = TRY(lookup_local_variable("IFS"sv));
+    if (auto const* value_from_env = getenv("IFS"); value_from_env)
+        split_by_any_of = TRY(String::from_utf8({ value_from_env, strlen(value_from_env) }));
+    else if (split_by_variable)
+        split_by_any_of = TRY(const_cast<AST::Value&>(*split_by_variable).resolve_as_string(*this));
+
+    auto file = TRY(Core::File::standard_input());
+    auto buffered_stream = TRY(Core::BufferedFile::create(move(file)));
+
+    StringBuilder builder;
+    ByteBuffer buffer;
+
+    enum class LineState {
+        Done,
+        EscapedNewline,
+    };
+    auto read_line = [&]() -> ErrorOr<LineState> {
+        if (m_is_interactive && isatty(STDIN_FILENO)) {
+            // Show prompt
+            warn("read: ");
+        }
+        size_t attempted_line_size = 32;
+
+        for (;;) {
+            auto result = buffered_stream->read_line(TRY(buffer.get_bytes_for_writing(attempted_line_size)));
+            if (result.is_error() && result.error().is_errno() && result.error().code() == EMSGSIZE) {
+                attempted_line_size *= 2;
```

Comment:
Kind of feels like Stream should have a read_line_buffer thing that does this for you. As-is, it looks a bit weird that this method that's about implementing `read` has to work at a different level of abstraction for getting a line of text into a buffer.

Reply:
I'd add something like that, but every other user has its own idea of how to read a line...and some people get sad when Stream grows an allocate-y method.
It does make sense to have one though